### PR TITLE
Input Wrappers: first pass

### PR DESCRIPTION
**Pull request by carlosscastro**
Includes:

- Text, Confirm, Integer and Float input wrappers
- Generic input wrapper
- Tests
- Samples
- Json
- Factory registration

Pending:

- Schemas
- Cheatsheet
- Choice prompt
- Datetime prompt
- Details on retry vs invalid. How much to defer to wrapped prompt vs customize

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/InputWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs.Rules.Input
+{
+    public class InputWrapper<TPrompt, TValue> : DialogCommand, IDialogDependencies where TPrompt : IDialog, new()
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/InputWrapper.cs]: This is the generic wrapper

File: libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        public TNumber MinValue { get; set; }
+        public TNumber MaxValue { get; set; }
+
+        public NumberInput()
+        {
+        }
+
+        protected override NumberPrompt<TNumber> CreatePrompt()
+        {
+            return new NumberPrompt<TNumber>(null, new PromptValidator<TNumber>(async (promptContext, cancel) =>
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs]: @tomlm is this what you had in mind when we talked?

**Comment by carlosscastro** [comment]
@tomlm first pass on inputs. Note that I didn't implement choice because it is a bit more involved and want to make sure the overall approach is fine.

File: libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs.Rules.Input
+{
+    public class NumberInput<TNumber> : InputWrapper<NumberPrompt<TNumber>, TNumber> where TNumber : struct, IComparable<TNumber>
+    {
+        public TNumber MinValue { get; set; }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs]: >MinValue [](start = 23, length = 8)

I'm not a huge fan of postfix "Value".  What's wrong with Min, Max ?

File: libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs (hunk 2 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Rules/Input/NumberInput.cs]: yep

---
In reply to: [269795826](https://github.com/Microsoft/botbuilder-dotnet/pull/1569#discussion_r269795826) [](ancestors = 269795826)

**Pull request merged**